_posts/2021-09-03-rudiments-of-data-wrangling-in-rust.md

```diff
@@ -0,0 +1,519 @@
+## Processing the CSV records
+
+Once we have our reader, we can start preparing to read the lines.
+From the reader we get an iterator which returns values of type
+`Result<csv::StringRecord, std::error::Error>`. Each `StringRecord`
+is essentially a vector of `String` values, representing all the
+fields. The delimiters and possible surrounding quotes are
+discarded.
+
+I'm interested in electric vehicle registrations of the last five
+years, from 2016 to 2020. This can be expressed as a Rust 
+[range](https://doc.rust-lang.org/reference/expressions/range-expr.html):
+
+    let year_range = 2016..=2020;
+
+I also need a counter that saves the number of EV registrations
+for each of these years. The `HashMap` struct in the Rust standard 
+library is suitable for this purpose, so let's make one, with the
+years as keys, and the counts as values:
+
+    let mut registrations = HashMap::<i32, u32>::new();
+
+This `HashMap` is mutable because we want to change its contents
+as we read in the lines from the CSV files. The default inferred
+type for integer literals in Rust is `u32`, but the type of the key
+is `i32` because that is also the type of the year in the date
+manipulation code found in the `chrono` crate, as you'll see soon.
+
+## Enumerated type detour
+
+While algebraic data types (as found in functional programming
+languages like [Haskell](https://www.haskell.org) and 
+[F#](https://fsharp.org)) are one of the distinctive features 
+of Rust, in this program I mainly used plain vanilla enumerated types
+to express the content of some of the data fields in the CSV file.
+
+I'm interested in the registrations of electric vehicles,
+and I was able to see from the Excel file accompanying the data
+that the driving power of a vehicle is encoded as a string with
+values like "01", "02", "03" etc. The value for electricity is "04".
+Rather than convert these into actual numbers, we can introduce
+an enumerated type:
+
+    enum DrivingPower {
+        Petrol,
+        DieselFuel,
+        FuelOil,
+        Electricity,
+        Hydrogen,
+        Gas,
+        Methanol,
+        BiodieselFuel,
+    }
+
+This type has a couple of more values than we need here (and this
+is not even exhaustive (pun not intended)), but this
+is for illustration and future use.
+
+To be able to use this enum while processing the CSV records,
+we define a function that can convert the symbolic values into the
+actual strings found in the CSV field:
+
+    impl DrivingPower {
+        fn as_str(&self) -> &'static str {
+            match *self {
+                DrivingPower::Petrol => "01",
+                DrivingPower::DieselFuel => "02",
+                DrivingPower::FuelOil => "03",
+                DrivingPower::Electricity => "04",
+                DrivingPower::Hydrogen => "05",
+                DrivingPower::Gas => "06",
+                DrivingPower::Methanol => "07",
+                DrivingPower::BiodieselFuel => "08",
+            }
+        }
+    }
```

Comment:
There often exists a standard library trait for these kinds of common use cases. [FromStr](https://doc.rust-lang.org/std/str/trait.FromStr.html) could be used here.